### PR TITLE
feat(auth): automatically renew an access_token before it expires

### LINKED ISSUE #44: Automatically renew access_token before it expires
See old task https://github.com/sphereio/sphere-node-connect/issues/7

### PR DESCRIPTION
**Pull request by emmenko**
- [x] commit messages are commitizen-friendly
- [x] code is unit tested
- [ ] code is integration tested
- [ ] public code is documented
- [x] code is reviewed by at least one person
- [x] code satisfies linter rules

fixes #44

/cc @PhilippSpo

### CODE REVIEW

File: src/coffee/connect/rest.coffee [hunk 1]
```diff
@@ -119,7 +119,15 @@ class Rest
   # Throws an {Error} if `access_token` could not be retrieved
   _preRequest: (params, callback) ->
     _req = (retry) =>
-      unless @_options.access_token
+      isTokenExpired = false
+      if @_tokenExpiration
+        expirationTime = Date.now() + (@_tokenExpiration * 1000) -
```
Comment (butenkor) [inline src/coffee/connect/rest.coffee]: Why do you use current date? You should have used here the date which has been used during last token creation, right?

Comment (emmenko) [inline src/coffee/connect/rest.coffee]: No, what you get from the API when you request an `access_token` is the `expires_in` value (2 days).
The `_tokenExpiration` is that value.

What I'm doing here is to calculate the **future** date, which consists in:

```
now + expirationPeriod - gap
```

Comment (butenkor) [inline src/coffee/connect/rest.coffee]: How can future date get ever older then now if you alwas take now + 2 days?

Comment (emmenko) [inline src/coffee/connect/rest.coffee]: Ah now I get what you mean :)

Yeah you're right, I've put it in the wrong place. Thanks for catching this! 👍

Comment (emmenko) [inline src/coffee/connect/rest.coffee]: There you go https://github.com/sphereio/sphere-node-sdk/pull/127/commits/ad1e2c6ffb069fbd820eda9d19f6f8fff57d2c26

**Comment by PhilippSpo** [comment]
👍

**Pull request merged**